File: java/Samples/Biometrics/Java/abis-sample/src/main/java/com/neurotec/samples/abis/swing/MatchingResultView.java
package com.neurotec.samples.abis.swing;

import com.neurotec.biometrics.NBiometricType;
import com.neurotec.biometrics.NEMatchingDetails;
import com.neurotec.biometrics.NFMatchingDetails;
import com.neurotec.biometrics.NLMatchingDetails;
import com.neurotec.biometrics.NMatchingDetails;
import com.neurotec.biometrics.NMatchingResult;
import com.neurotec.biometrics.NSMatchingDetails;

import java.awt.Color;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.util.ArrayList;
import java.util.List;

import javax.swing.BorderFactory;
import javax.swing.BoxLayout;
import javax.swing.JLabel;
import javax.swing.JPanel;

public class MatchingResultView extends JPanel implements ActionListener {

	private static final long serialVersionUID = 1L;

	private final NMatchingResult matchingResult;
	private final int matchingThreshold;
	private final List<ActionListener> actionListeners;

	private boolean linkEnabled = true;
	private JLabel lblDetails;
	private LinkButton btnLink;

	public MatchingResultView(NMatchingResult matchingResult, int matchingThreshold) {
		this.matchingResult = matchingResult;
		this.matchingThreshold = matchingThreshold;
		this.actionListeners = new ArrayList<ActionListener>();
		initGUI();
		btnLink.setText(String.format("<html><color=\"black\"><a href=\"\">Matched with %s, score = %d</a></color></html>", this.matchingResult.getId(), this.matchingResult.getScore()));
		lblDetails.setText(matchingResultToString(this.matchingResult));
	}

	private void initGUI() {
		btnLink = new LinkButton();
		lblDetails = new JLabel();

		setLayout(new BoxLayout(this, BoxLayout.Y_AXIS));

		btnLink.setBorder(BorderFactory.createEmptyBorder(3, 3, 3, 3));
		add(btnLink);

		lblDetails.setBorder(BorderFactory.createEmptyBorder(3, 10, 3, 3));
		add(lblDetails);

		btnLink.addActionListener(this);
	}

	private String matchingResultToString(NMatchingResult result) {
		NMatchingDetails details = result.getMatchingDetails();
		StringBuilder sb = new StringBuilder(256);
		sb.append("<html>");
		if (details != null) {
			String strBelowThreshold = " (Below matching threshold)";
			if (details.getBiometricType().contains(NBiometricType.FACE)) {
				sb.append(String.format("Face match details: score = %d<br/>", details.getFacesScore()));
				int index = 0;
				for (NLMatchingDetails faceDetails : details.getFaces()) {
					if (faceDetails.getMatchedIndex() == -1) {
						sb.append(String.format("Face index %d: doesn't match<br/>", index++));
					} else {
						sb.append(String.format("Face index %d: matched with index %d, score = %d%s;<br/>", index++, faceDetails.getMatchedIndex(), faceDetails.getScore(), faceDetails.getScore() < matchingThreshold ? strBelowThreshold : ""));
					}
				}
			}
			if (details.getBiometricType().contains(NBiometricType.FINGER)) {
				sb.append(String.format("Finger match details: score = %d<br/>", details.getFingersScore()));
				int index = 0;
				for (NFMatchingDetails fingerDetails : details.getFingers()) {
					if (fingerDetails.getMatchedIndex() == -1) {
						sb.append(String.format("Finger index %d: doesn't match<br/>", index++));
					} else {
						sb.append(String.format("Finger index %d: matched with index %d, score = %d%s;<br/>", index++, fingerDetails.getMatchedIndex(), fingerDetails.getScore(), fingerDetails.getScore() < matchingThreshold ? strBelowThreshold : ""));
					}
				}
			}
			if (details.getBiometricType().contains(NBiometricType.IRIS)) {
				sb.append(String.format("Iris match details: score = %d<br/>", details.getIrisesScore()));
				int index = 0;
				for (NEMatchingDetails irisDetails : details.getIrises()) {
					if (irisDetails.getMatchedIndex() == -1) {
						sb.append(String.format("Iris index %d: doesn't match<br/>", index++));
					} else {
						sb.append(String.format("Iris index %d: matched with index %d, score = %d%s;<br/>", index++, irisDetails.getMatchedIndex(), irisDetails.getScore(), irisDetails.getScore() < matchingThreshold ? strBelowThreshold : ""));
					}
				}
			}
			if (details.getBiometricType().contains(NBiometricType.PALM)) {
				sb.append(String.format("Palm match details: score = %d<br/>", details.getPalmsScore()));
				int index = 0;
				for (NFMatchingDetails palmDetails : details.getPalms()) {
					if (palmDetails.getMatchedIndex() == -1) {
						sb.append(String.format("Palm index %d: doesn't match<br/>", index++));
					} else {
						sb.append(String.format("Palm index %d: matched with index %d, score = %d%s;<br/>", index++, palmDetails.getMatchedIndex(), palmDetails.getScore(), palmDetails.getScore() < matchingThreshold ? strBelowThreshold : ""));
					}
				}
			}
			if (details.getBiometricType().contains(NBiometricType.VOICE)) {
				sb.append(String.format("Voice match details: score = %d<br/>", details.getVoicesScore()));
				int index = 0;
				for (NSMatchingDetails voiceDetails : details.getVoices()) {
					if (voiceDetails.getMatchedIndex() == -1) {
						sb.append(String.format("Voice index %d: doesn't match<br/>", index++));
					} else {
						sb.append(String.format("Voice index %d: matched with index %d, score = %d%s;<br/>", index++, voiceDetails.getMatchedIndex(), voiceDetails.getScore(), voiceDetails.getScore() < matchingThreshold ? strBelowThreshold : ""));
					}
				}
			}
			if (!details.getBiometricType().contains(NBiometricType.FACE)
				&& !details.getBiometricType().contains(NBiometricType.FINGER)
				&& !details.getBiometricType().contains(NBiometricType.IRIS)
				&& !details.getBiometricType().contains(NBiometricType.PALM)
				&& !details.getBiometricType().contains(NBiometricType.VOICE)) {
				sb.append("Score = ").append(details.getScore());
			}
		}
		return sb.append("</html>").toString();
	}

	public void addActionListener(ActionListener l) {
		actionListeners.add(l);
	}

	public void removeActionListener(ActionListener l) {
		actionListeners.remove(l);
	}

	public boolean isLinkEnabled() {
		return linkEnabled;
	}

	public void setLinkEnabled(boolean value) {
		this.linkEnabled = value;
		btnLink.setEnabled(value);
		btnLink.setForeground(value ? Color.GREEN : Color.BLACK);
	}

	public NMatchingResult getMatchingResult() {
		return matchingResult;
	}

	@Override
	public void actionPerformed(ActionEvent ev) {
		if (ev.getSource().equals(btnLink)) {
			for (ActionListener l : actionListeners) {
				l.actionPerformed(new ActionEvent(this, ActionEvent.ACTION_PERFORMED, "LinkClicked"));
			}
		}
	}

}

File: java/Samples/Biometrics/CPP/AbisSampleWX/Common/MatchingResultView.cpp
#include "Precompiled.h"

#include <Common/MatchingResultView.h>

using namespace Neurotec::Biometrics;

namespace Neurotec { namespace Samples
{

MatchingResultView::MatchingResultView(wxWindow *parent, wxWindowID winid) :
	wxPanel(parent, winid),
	m_matchingResult(NULL),
	m_matchingThreshold(0)
{
	m_linkEnabled = false;
	m_linkPressedCallback = NULL;
	m_callbackParam = NULL;
	m_isLinkActive = false;

	CreateGuiElements();
	RegisterGuiEvents();
}

MatchingResultView::~MatchingResultView()
{
	UnregisterGuiEvents();
}

void MatchingResultView::SetMatchingThreshold(int value)
{
	m_matchingThreshold = value;
}

void MatchingResultView::SetIsLinkActive(bool value)
{
	m_isLinkActive = value;
}

bool MatchingResultView::IsLinkActive()
{
	return m_isLinkActive;
}

void MatchingResultView::SetMatchingResult(NMatchingResult matchingResult)
{
	m_matchingResult = matchingResult;

	if (!matchingResult.IsNull())
	{
		m_lblLink->SetLabelText(wxString::Format(wxT("Matched with '%s', score = %d"), (wxString)m_matchingResult.GetId(), m_matchingResult.GetScore()));
		m_lblDetails->SetLabelText(MatchingResultToString(m_matchingResult));
	}

	this->Layout();
}

wxString MatchingResultView::MatchingResultToString(const ::Neurotec::Biometrics::NMatchingResult matchingResult)
{
	NMatchingDetails details = matchingResult.GetMatchingDetails();

	wxString result = wxEmptyString;

	if (details.IsNull())
	{
		return result;
	}

	int index;
	int score;
	wxString bellowThreshold = wxT(" (Below matching threshold)");
	if ((details.GetBiometricType() & nbtFace) == nbtFace)
	{
		result += wxString::Format(wxT("Face match details: score = %d\n"), details.GetFacesScore());

		NMatchingDetails::FaceCollection faces = details.GetFaces();
		for (int i = 0; i < faces.GetCount(); i++)
		{
			NLMatchingDetails faceDetails = faces[i];
			index = faceDetails.GetMatchedIndex();
			if (index != -1)
			{
				score = faceDetails.GetScore();
				result += wxString::Format(wxT("    face index %d: matched with index %d, score = %d%s\n"), i, index, score, score < m_matchingThreshold ? bellowThreshold : (wxString)wxEmptyString);
			}
			else
			{
				result += wxString::Format(wxT("    face index %d: doesn't match\n"), i);
			}
		}
	}

	if ((details.GetBiometricType() & nbtFinger) == nbtFinger)
	{
		result += wxString::Format(wxT("Fingerprint match details: score = %d\n"), details.GetFingersScore());

		NMatchingDetails::FingerCollection fingers = details.GetFingers();
		for (int i = 0; i < fingers.GetCount(); i++)
		{
			NFMatchingDetails fingerDetails = fingers[i];
			index = fingerDetails.GetMatchedIndex();
			if (index != -1)
			{
				score = fingerDetails.GetScore();
				result += wxString::Format(wxT("    fingerprint index %d: matched with index %d, score = %d%s\n"), i, index, score, score < m_matchingThreshold ? bellowThreshold : (wxString)wxEmptyString);
			}
			else
			{
				result += wxString::Format(wxT("    fingerprint index %d: doesn't match\n"), i);
			}
		}
	}

	if ((details.GetBiometricType() & nbtIris) == nbtIris)
	{
		result += wxString::Format(wxT("Irises match details: score = %d\n"), details.GetIrisesScore());

		NMatchingDetails::IrisCollection irises = details.GetIrises();
		for (int i = 0; i < irises.GetCount(); i++)
		{
			NEMatchingDetails irisesDetails = irises[i];
			index = irisesDetails.GetMatchedIndex();
			if (index != -1)
			{
				score = irisesDetails.GetScore();
				result += wxString::Format(wxT("    irises index %d: matched with index %d, score = %d%s\n"), i, index, score, score < m_matchingThreshold ? bellowThreshold : (wxString)wxEmptyString);
			}
			else
			{
				result += wxString::Format(wxT("    irises index %d: doesn't match\n"), i);
			}
		}
	}

	if ((details.GetBiometricType() & nbtPalm) == nbtPalm)
	{
		result += wxString::Format(wxT("Palmprint match details: score = %d\n"), details.GetPalmsScore());

		NMatchingDetails::PalmCollection palms = details.GetPalms();
		for (int i = 0; i < palms.GetCount(); i++)
		{
			NFMatchingDetails palmsDetails = palms[i];
			index = palmsDetails.GetMatchedIndex();
			if (index != -1)
			{
				score = palmsDetails.GetScore();
				result += wxString::Format(wxT("    palmprint index %d: matched with index %d, score = %d%s\n"), i, index, score, score < m_matchingThreshold ? bellowThreshold : (wxString)wxEmptyString);
			}
			else
			{
				result += wxString::Format(wxT("    palmprint index %d: doesn't match\n"), i);
			}
		}
	}

	if ((details.GetBiometricType() & nbtVoice) == nbtVoice)
	{
		result += wxString::Format(wxT("Voice match details: score = %d\n"), details.GetVoicesScore());

		NMatchingDetails::VoiceCollection voices = details.GetVoices();
		for (int i = 0; i < voices.GetCount(); i++)
		{
			NSMatchingDetails voiceDetails = voices[i];
			index = voiceDetails.GetMatchedIndex();
			if (index != -1)
			{
				score = voiceDetails.GetScore();
				result += wxString::Format(wxT("    voices index %d: matched with index %d, score = %d%s\n"), i, index, score, score < m_matchingThreshold ? bellowThreshold : (wxString)wxEmptyString);
			}
			else
			{
				result += wxString::Format(wxT("    voices index %d: doesn't match\n"), i);
			}
		}
	}

	if ((details.GetBiometricType() & (nbtFinger | nbtPalm | nbtIris | nbtVoice | nbtFace)) == nbtNone)
	{
		result += wxString::Format(wxT(" score = %d"), details.GetScore());
	}

	return result;
}

void MatchingResultView::OnHyperlinkClick(wxHyperlinkEvent&)
{
	if (!m_isLinkActive)
	{
		return;
	}

	if (m_linkPressedCallback != NULL)
	{
		(*m_linkPressedCallback)(m_matchingResult, m_callbackParam);
	}
}

void MatchingResultView::SetLinkPressedCallback(MatchingResultView::LinkPressedCallback callback, void *param)
{
	m_linkPressedCallback = callback;
	m_callbackParam = param;
}

void MatchingResultView::RegisterGuiEvents()
{
	m_lblLink->Connect(wxEVT_COMMAND_HYPERLINK, wxHyperlinkEventHandler(MatchingResultView::OnHyperlinkClick), NULL, this);
}

void MatchingResultView::UnregisterGuiEvents()
{
	m_lblLink->Disconnect(wxEVT_COMMAND_HYPERLINK, wxHyperlinkEventHandler(MatchingResultView::OnHyperlinkClick), NULL, this);
}

void MatchingResultView::CreateGuiElements()
{
	wxBoxSizer *sizer = new wxBoxSizer(wxVERTICAL);
	this->SetSizer(sizer);

	m_lblLink = new wxHyperlinkCtrl(this, wxID_ANY, wxEmptyString, wxT("#"), wxDefaultPosition, wxDefaultSize, wxHL_DEFAULT_STYLE);
	m_lblLink->SetToolTip(wxEmptyString);
	sizer->Add(m_lblLink, 0, wxALL);

	m_lblDetails = new wxStaticText(this, wxID_ANY, wxEmptyString);
	sizer->Add(m_lblDetails, 0, wxLEFT, 10);

	this->Layout();
}

}}

